### ISSUE TITLE
Explicitly control which map tiles get loaded

### ISSUE THREAD

**Issue by chrisdalke** [felixpalmer/procedural-gl-js#19]
Hello,

Thanks for your great work on this library!

I'm evaluating using procedural-gl-js in a visualization tool, to plot 3D GPS flight paths. In this application, the viewer is only concerned with the area immediately around the flight path. This flight path might be small (a few kilometers), large (hundreds of kilometers), and irregularly shaped (a random path that stretches very far in some direction).

**Is there any way to configure the range of tiles loaded around the display location, or explicitly choose which tiles to load?** 
Right now, the map seems to load a pretty large area in a square around the GPS coordinates specified using displayLocation. For example, if I search "Boston, MA", the example loads almost all of Massachusetts. I'm hoping to do a few things:

1) Only load the map tiles directly underneath the flight path
2) Load tiles that may stretch very far in one direction, perhaps horizontally across 100s of kilometers.

I'd love to be able to pass in an explicit list of XYZ tile coordinates,. I looked around the source code, but couldn't find where the library is converting lat/lng coordinates into an list of tiles to load. Would be happy to contribute if I could be pointed in the right direction!

**Comment by chrisdalke** [comment]
To further clarify, my idea was that you could perhaps pass in an option on `Procedural.displayLocation` that would set the range of tiles loaded, or even explicitly define the list of tiles.

**Comment by felixpalmer** [comment]
I'm not really clear on what you want to achieve, when you're plotting a flight path surely you want the distant terrain to still be visible? It is true the map loads a large area around the current current position, but the LOD system means that the landscape far away is cover by low resolution data so is not needlessly wasting resources and bandwidth. If it were possible to load a list of XYZ tile like you describe then you would end up with blank parts of your terrain, no?

All this said, I have been thinking about extending the way a datasource is defined so that the `urlFormat` parameter accepts a function as well as a template string, which would allow a dynamic way of loading in tiles (e.g. one data source for high zoom, another for low) and in your case you could only return a valid tile url for the areas you wanted, thus only loading the tiles you need

**Comment by chrisdalke** [comment]
> when you're plotting a flight path surely you want the distant terrain to still be visible?

True, however the camera perspective means the majority of these tiles aren't visible unless you scroll very far to view them.

Here's an example -- I've drawn a line from Boston to New York City (~300km). The line renders fine, but at a certain point it keeps going past where the map loads. What I'm really looking for is to configure the default range to a much lower value, and stream a new set of tiles when the camera location changes.

My original thought was that exposing the "tile picking" algorithm could allow control over which tiles load, but that's probably too low-level. A better phrased question: Do you have plans to stream new tiles when the user moves their camera to the edge of the loaded terrain?

![procedural-js-line-1](https://user-images.githubusercontent.com/11547438/105205175-f30a2780-5b12-11eb-9ae3-822506b96b54.PNG)
![procedural-js-line-2](https://user-images.githubusercontent.com/11547438/105205180-f43b5480-5b12-11eb-8f3c-9195b84bab95.PNG)

**Comment by felixpalmer** [comment]
TLDR: grab the new version (1.0.9) and you'll be able to move between New York and Boston without issue.

Your issue was a bit of a red herring. The problem was that the skydome was accidentally set to be static, was thus cutting off the map. Beyond this the tiles were loading, you just couldn't see them. I've now fixed this, along with tweaking the controls to allow movement over larger distances.

The map does load infinitely into the distance, and tiles are automatically appended on the edge as you approach it, being taken from the edge you can't see. The distance fog hides this process, giving the illusion of an infinite terrain.

Also, to clarify, the engine will only load higher resolution tiles in the area that are visible on screen, and the resolution is linked to the distance of the terrain to the camera.

Let me know if this resolves the issue for you. Looking forward to seeing what you build, sounds like a cool project. I'd be happy to share it when it is ready if you like.

**Comment by chrisdalke** [comment]
Oh awesome, I've updated and it works great now. That's funny, I saw the static skybox on the examples as well and assumed you hadn't implemented infinite scrolling yet. Thanks for the clearification!

I'll definitely let you know if this gets used on an application in production!
